drv/lpc55-sha256/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+//! results for the subset of test vectors in RFC 4231 that are consistent with
+//! the system's limitations (below). (As well as some manually generated test
+//! cases.) The stage0 codebase contains a test suite for its version of this
+//! code; because the code isn't currently shared, it's always possible they've
+//! diverged.
+//!
+//! # Assumptions and limitations
+//!
+//! This is a _deliberately_ very limited implementation, to save space and
+//! eliminate corner cases we'd otherwise have to think about.
+//!
+//! First, the SHA256 implementation:
+//!
+//! - Input is in units of 32-bit words only. Because of how SHA padding works,
+//!   this means you can't compute the correct SHA256 for (say) a 3-byte string.
+//!   However, the types used in these APIs will also prevent you from
+//!   _attempting_ to compute that SHA256, so this seems ok.
+//!
+//! - We're doing SHA256 _only,_ so block and hash sizes are fixed by compile
+//!   time constants, and input and output is in fixed-length arrays defined by
+//!   those constants.
+//!
+//! HMAC:
+//!
+//! - We only implement HMAC-SHA256 without truncation. (Truncation would be
+//!   easy to add outside this module if you want it.)
+//!
+//! - Keys are always exactly one SHA256 digest in length, which is to say, 256
+//!   bits. This eliminates some corner cases and conditionals from the key
+//!   processing code, not to mention a bunch of bounds checks.
+//!
+//! - Keys are delivered in blocks of 32-bit words to ensure alignment
+//!   (specifically, as fixed-length arrays of `u32`).
+//!
+//! - Data is also delivered in streams of 32-bit words, but can be any length.
+//!
+//! - For implementation convenience, data must be a single contiguous block of
+//!   32-bit words that can be addressed by one slice, which in theory limits
+//!   the maximum count to `isize::MAX`. In practice address space limitations
+//!   on our processor limit it to less than that (about 1 GiB at a time).
+//!   (This restriction would be straightforward to lift if required.)
+
+use core::num::Wrapping;
+use userlib::{sys_irq_control, sys_recv_closed, TaskId};
+
+// These constants describe intrinsic properties of the SHA256 algorithm and
+// should not be changed.
+const WORDS_PER_BLOCK: usize = 512 / 32; // which is to say, 16
+const WORDS_PER_HASH: usize = 256 / 32; // which is to say, 8
+
+// It's also convenient to have one as Wrapping u64:
+const WORDS_PER_BLOCK64: Wrapping<u64> = Wrapping(WORDS_PER_BLOCK as u64);
+
+/// Computes an HMAC-SHA256, a kind of keyed hash, over `data` using `key`.
+///
+/// The `key` is fixed at 256 bits, and both `key` and `data` must consist of a
+/// whole number of 32-bit words. See the module docs for more information about
+/// these limitations.
+///
+/// HMAC is specified in RFC 2104.
+///
+/// `notification_mask` is the notification mask corresponding to the HASHCRYPT
+/// IRQ in your task. If you'd rather busywait, pass 0.
+#[inline(never)]
```

Comment:
Should this docstring have the same warning about resetting the hash unit, i.e.
```rust
    /// You are _strongly advised_ to have reset the HASHCRYPT unit just before
    /// this, lest it contain nonsense left over from a previous hash. (In
    /// particular, the boot ROM leaves nonsense in it, so it needs to be reset
    /// at least once after startup.)
```
(or equivalent)?

Reply:
......so, actually, because of the reset behavior, this hmac function is _incorrect_ in this context. (It needs to be able to reset the unit between the two constituent hashes, and doesn't know how.)

I could fix that by having the caller pass in a reset thunk, but I'll just remove it for now.